### PR TITLE
New Graphic File Format Repo Update + Bug Fixes

### PR DESCRIPTION
**Pull request by dustin4242**

### CODE REVIEW

File: src/graphics.rs [hunk 1]
```diff
@@ -8,31 +8,42 @@ pub struct Graphics {
...
     let mut data: [[u8; 16]; SCREEN_WIDTH / 8 * SCREEN_HEIGHT / 16] =
         [[0; 16]; SCREEN_WIDTH / 8 * SCREEN_HEIGHT / 16];
     for y in 0..height {
         for x in 0..width {
             let mut character: [u8; 16] = [0; 16];
             for h in 0..16 {
                 character[h] = *graphical_data
-                    .get((y * width * 16) as usize + (h * width as usize) + x as usize)
-                    .unwrap_or(&0x00);
+                    .get((y * width * 16) as usize + (h * width as usize))
```
Comment (dustin4242) [inline src/graphics.rs:26]: Fixes issues on "multiple" graphics being printed at once

File: src/graphics.rs [hunk 2]
```diff
@@ -8,31 +8,42 @@ pub struct Graphics {
...
+            let mut rgb = (0, 0, 0);
+            rgb.0 = *graphic.get(color_index + x * 3).unwrap();
+            rgb.1 = *graphic.get(color_index + x * 3 + 1).unwrap();
+            rgb.2 = *graphic.get(color_index + x * 3 + 2).unwrap();
+            temp[x] = rgb;
+        }
+        color_pallete = Some(temp);
+    }
+    color_index += 8 * 3;
+    let color_data = graphic.get(color_index..);
```
Comment (dustin4242) [inline src/graphics.rs:46]: Better positioning for this since we can have graphics without color

File: src/screen.rs [hunk 3]
```diff
@@ -46,6 +47,15 @@ impl Screen {
...
         let mut color_index = 0;
+        unsafe { Port::new(0x3C8).write(0xF8 as u8) };
+        let mut data_port = Port::new(0x3C9);
+        for i in graphics.color_pallete.unwrap_or([(63, 63, 63); 8]) {
+            unsafe {
+                data_port.write(i.0);
+                data_port.write(i.1);
+                data_port.write(i.2);
+            }
+        }
```
Comment (dustin4242) [inline src/screen.rs:58]: Sets the 8 colors at the end of the vga palette with graphic colors

File: src/screen.rs [hunk 4]
```diff
@@ -55,29 +65,23 @@ impl Screen {
...
-                                )
-                                .unwrap_or(&0x0F);
+                            let color = 0xF8
+                                + ((*graphics
+                                    .color_data
+                                    .unwrap_or(&[])
+                                    .get(color_index / 2)
+                                    .unwrap_or(&0)
+                                    & (0x03 << (color_index % 2) * 4))
+                                    >> ((color_index % 2) * 4));
```
Comment (dustin4242) [inline src/screen.rs:82]: Honestly I like how small this is now

File: src/screen.rs [hunk 5]
```diff
@@ -55,29 +65,23 @@ impl Screen {
                     for x2 in 0..8 {
                         buffer.chars[self.row * 16 + h][self.column * 8 + x2] = if ((graphics
                             .data
-                            .get((y * SCREEN_WIDTH as u16 / 8 + x) as usize)
+                            .get((y * graphics.width + x) as usize)
```
Comment (dustin4242) [inline src/screen.rs:68]: Fixes issue when printing a graphic with height more than one

File: src/graphics.rs [hunk 6]
```diff
@@ -8,31 +8,42 @@ pub struct Graphics {
...
+                    .get((y * width * 16) as usize + (h * width as usize))
+                    .unwrap();
             }
             data[(y * width + x) as usize] = character;
         }
     }
+    let mut color_index = (width * height * 16) as usize + 2;
+    let mut color_pallete = None;
+    if graphic.get(color_index..color_index + 8 * 3).is_some() {
+        let mut temp: [(u8, u8, u8); 8] = [(0, 0, 0); 8];
```
Comment (talwat) [inline src/graphics.rs:35]: The name "temp" isn't very descriptive. Could you come up with something else, or move the whole contents of the loop to a separate function.

File: src/screen.rs [hunk 7]
```diff
@@ -46,6 +47,15 @@ impl Screen {
     pub fn print_graphics(&mut self, graphics: Graphics) {
         let buffer = unsafe { self.buffer.as_mut().unwrap() };
         let mut color_index = 0;
+        unsafe { Port::new(0x3C8).write(0xF8 as u8) };
+        let mut data_port = Port::new(0x3C9);
```
Comment (talwat) [inline src/screen.rs:51]: Maybe a comment on what this actually is would help.

**Review by talwat — APPROVED** [review]
Just a few small things that could be edited. LGTM :)

File: src/graphics.rs (hunk 6 above)
Comment (dustin4242) [reply src/graphics.rs:35]: agreed

File: src/screen.rs (hunk 7 above)
Comment (dustin4242) [reply src/screen.rs:49]: agreed

**Pull request merged**